esp-link/cgiwifi.c

```diff
@@ -281,239 +298,427 @@ int ICACHE_FLASH_ATTR cgiWiFiScan(HttpdConnData *connData) {
+        if (gl > 0) ok = ok && parse_ip(gateway, &ipi.gw);
+        else ipi.gw.addr = 0;
+        if (!ok) {
+            jsonHeader(connData, 400);
+            httpdSend(connData, "Cannot parse static IP config", -1);
+            return HTTPD_CGI_DONE;
+        }
+        // save the params in flash
+        flashConfig.staticip = ipi.ip.addr;
+        flashConfig.netmask = ipi.netmask.addr;
+        flashConfig.gateway = ipi.gw.addr;
+        // construct redirect URL
+        os_sprintf(url, "{\"url\": \"http://%d.%d.%d.%d\"}", IP2STR(&ipi.ip));
+        
+    } else {
+        // dynamic IP
+        flashConfig.staticip = 0;
+        os_sprintf(url, "{\"url\": \"http://%s\"}", flashConfig.hostname);
+    }
+    
+    configSave(); // ignore error...
+    // schedule change-over
+    os_timer_disarm(&reassTimer);
+    os_timer_setfn(&reassTimer, configWifiIP, NULL);
+    os_timer_arm(&reassTimer, 1000, 0); // 1 second for the response of this request to make it
+    // return redirect info
+    jsonHeader(connData, 200);
+    httpdSend(connData, url, -1);
+    return HTTPD_CGI_DONE;
+}
 
-  // get args and their string lengths
-  int dl = httpdFindArg(connData->getArgs, "dhcp", dhcp, sizeof(dhcp));
-  int sl = httpdFindArg(connData->getArgs, "staticip", staticip, sizeof(staticip));
-  int nl = httpdFindArg(connData->getArgs, "netmask", netmask, sizeof(netmask));
-  int gl = httpdFindArg(connData->getArgs, "gateway", gateway, sizeof(gateway));
+// ==== Soft-AP related functions
+
+// Change Soft-AP main settings
+int ICACHE_FLASH_ATTR cgiApSettingsChange(HttpdConnData *connData) {
+    
+    if (connData->conn==NULL) return HTTPD_CGI_DONE; // Connection aborted. Clean up.
+    
+    
+    // No changes for Soft-AP in STA mode
+    int mode = wifi_get_opmode();
+    if ( mode == 1 ){
+        jsonHeader(connData, 400);
+        httpdSend(connData, "No changes allowed in STA mode", -1);
+        return HTTPD_CGI_DONE;
+    }
+    // Get Soft-Ap config, just in case
+    wifi_softap_get_config(&apconf);
+    
+    char buff[96];
+    int len;
+    // Do we need a password?
+    int pass_need=1;
+    
+    // Check extra security measure
+    len=httpdFindArg(connData->getArgs, "100", buff, sizeof(buff));
+    if(len>0){
+        if(atoi(buff)!=1){
+            jsonHeader(connData, 400);
+            return HTTPD_CGI_DONE;
+        }
+    }
+    // Get the new SSID and set
+    len=httpdFindArg(connData->getArgs, "ap_ssid", buff, sizeof(buff));
+    if(checkString(buff) && len>7 && len<32){
+        // STRING PREPROCESSING DONE IN CLIENT SIDE
+        memset(apconf.ssid, 0, 32);
+        os_memcpy(apconf.ssid, buff, len);
+        apconf.ssid_len = len;
+        pass_need = 1;
+    }else{
+        pass_need = 0;
+        jsonHeader(connData, 400);
+        httpdSend(connData, "SSID not valid or out of range", -1);
+        return HTTPD_CGI_DONE;
+    }
+    // Set new PASSWORD
+    if( pass_need ){
```

Comment:
I did't want to create several char array to store each parameter and call httpdFindArg only when needed, I used pass_need to know if I need to get further looking for password and auth mode arguments, depending on ssid value and the password value itself. 
This way if ssid or password is invalid or out of range, it would save some memory and cpu time.